Mathlib/Order/KrullDimension.lean

```diff
@@ -32,67 +29,13 @@ sum of `-∞` and the Krull dimension of any other varieties.
 variable (α : Type _) [Preorder α]
 
 /--
-For a preordered set `(α, <)`, a strict series of `α` of length `n` is a strictly monotonic function
-`Fin (n + 1) → α`, i.e. `a₀ < a₁ < ... < aₙ` with `aᵢ : α`.
--/
-structure StrictSeries where
-  /-- the number of inequalities in the series -/
-  length : ℕ
-  /-- the underlying function of a strict series -/
-  toFun : Fin (length + 1) → α
-  /-- the underlying function should be strictly monotonic -/
-  StrictMono : StrictMono toFun
-
-namespace StrictSeries
-
-instance : CoeFun (StrictSeries α) (fun x ↦ Fin (x.length + 1) → α) :=
-  { coe := StrictSeries.toFun }
-
-instance : Preorder (StrictSeries α) :=
-  Preorder.lift StrictSeries.length
-
-variable {α}
-
-lemma le_def (x y : StrictSeries α) : x ≤ y ↔ x.length ≤ y.length :=
-  Iff.rfl
-
-lemma lt_def (x y : StrictSeries α) : x < y ↔ x.length < y.length :=
-  Iff.rfl
-
-/--
-In a preordered set `α`, each term of `α` gives a strict series with the right most index to be 0.
--/
-@[simps!] def singleton (a : α) : StrictSeries α :=
-  { length := 0
-    toFun := fun _ ↦ a
-    StrictMono := fun _ _ h ↦ (ne_of_lt h $ @Subsingleton.elim _ subsingleton_fin_one _ _).elim }
-
-instance [IsEmpty α] : IsEmpty (StrictSeries α) :=
-  { false := fun x ↦ IsEmpty.false (x 0) }
-
-instance [Inhabited α] : Inhabited (StrictSeries α) :=
-  { default := singleton default }
-
-instance [Nonempty α] : Nonempty (StrictSeries α) :=
-  Nonempty.map singleton inferInstance
-
-lemma top_len_unique [OrderTop (StrictSeries α)] (p : StrictSeries α) (hp : IsTop p) :
-    p.length = (⊤ : StrictSeries α).length :=
-  le_antisymm (@le_top (StrictSeries α) _ _ _) (hp ⊤)
-
-lemma top_len_unique' (H1 H2 : OrderTop (StrictSeries α)) : H1.top.length = H2.top.length :=
-  le_antisymm (H2.le_top H1.top) (H1.le_top H2.top)
-
-end StrictSeries
-
-/--
-Krull dimension of a preordered set `α` is the supremum of the right most index of all strict
-series of `α`. If there is no strict series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull dimension
-is defined to be negative infinity; if the length of `a₀ < a₁ < ... < aₙ` is unbounded, its Krull
-dimension is defined to be positive infinity.
+Krull dimension of a preordered set `α` is the supremum of the right most index of all relation
+series of `α` order by `<`. If there is no series `a₀ < a₁ < ... < aₙ` in `α`, then its Krull
+dimension is defined to be negative infinity; if the length of all series `a₀ < a₁ < ... < aₙ` is
+unbounded, its Krull dimension is defined to be positive infinity.
 -/
 noncomputable def krullDim : WithBot (WithTop ℕ) :=
-  ⨆ (p : StrictSeries α), p.length
+  ⨆ (p : LTSeries α), p.length
```

Comment:
This is basically `Order.pred ((Set.univ : Set α) .chain_height : WithBot ℕ∞)` (once generalised to an arbitrary relation).

Reply:
I think it's fundamentally not a pred_order

Reply:
@YaelDillies do you think changing `s.chainHeight` to be `withBot ℕ∞` a good idea?

Reply:
Discussions on Zulip seems to conclude it's nicer to use `Fin` indexed tuples rather than `List`, and `RelSeries` rather than `List.chain'`. I think one reason it's nicer is that we restrict to positive length tuples and have direct access to the first and last terms of the chain without worrying about the `List.nil` case. Of course we may bundle the `≠ List.nil` condition or use a pair `(a,l)` as in `List.chain` to form the `RelSeries` type, but why not just generalize `CompositionSeries` given that there's already API around it. For example, `snoc` in #3852 and `append` in #3858 are already there, and now that we've `StrictMono` with the `step` condition we can simplify the proofs by copying from `CompositionSeries`. Worth noting that `append` in #3858 has length `p.length + q.length + 1` rather than `p.length + q.length` as in `CompositionSeries`. I think the latter is usually what we want. (For `List` there are List.Chain'.append and [List.Chain'.append_overlap](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Data/List/Chain.html#List.Chain'.append_overlap).)

Using `RelSeries` instead of `List.chain'` would probably mean abandonment of the `Set.chainHeight` API, which is unfortunate, and I wonder whether @erdOne has anything to say about it. `Set.chainHeight` is "off by one" and It's probably not worth redefining it if we're not gonna use it, but we may use an OrderIso between the two to transfer results if we want.

Now, some comments about the name `krullDim`: since you'd like to use it to define length of modules, I'd suggest calling it some generic name like `supSeriesLength` or `supChainLength`, and reserving `krullDim` for `CommRing.krullDim` or `TopologicalSpace.krullDim`. You won't want it to clash with [Krull dimension of modules](https://en.wikipedia.org/wiki/Deviation_of_a_poset) which is distinct from the length and defined in a totally different way.

Reply:
Oh, I just noticed #5881 already generalized `chainHeight` to arbitrary relations. Maybe we should just define `pred: ℕ∞ ≃o ℕ±∞`, prove that `pred set.univ.chainHeight = krullDim α` and `pred s.chainHeight = krullDim ↥s`, and start transferring result from `chainHeight` to `krullDim` ...

Reply:
> Now, some comments about the name krullDim: since you'd like to use it to define length of modules, I'd suggest calling it some generic name like supSeriesLength or supChainLength, and reserving krullDim for CommRing.krullDim or TopologicalSpace.krullDim. You won't want it to clash with [Krull dimension of modules](https://en.wikipedia.org/wiki/Deviation_of_a_poset) which is distinct from the length and defined in a totally different way.

Thanks for pointing this out

Reply:
> in #6277, I made the other way round, I rewrite `JordanHolder.lean` completely in terms of `RelSeries`

Yeah, that's exactly what I mean and I think #6277 is in the right direction. Fantastic work!